fvm/environment/event_emitter.go

```diff
@@ -0,0 +1,211 @@
+package environment
+
+import (
+	"fmt"
+
+	"github.com/onflow/cadence"
+	jsoncdc "github.com/onflow/cadence/encoding/json"
+
+	"github.com/onflow/flow-go/fvm/errors"
+	metering "github.com/onflow/flow-go/fvm/meter"
+	"github.com/onflow/flow-go/fvm/systemcontracts"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/trace"
+)
+
+// EventEmitter collect events, separates out service events, and enforces
+// event size limits.
+//
+// Note that scripts do not emit events, but must expose the API in compliance
+// with the runtime environment interface.
+type EventEmitter interface {
+	// Cadence's runtime API.  Note that the script variant will return
+	// OperationNotSupportedError.
+	EmitEvent(event cadence.Event) error
+
+	Events() []flow.Event
+	ServiceEvents() []flow.Event
+}
+
+var _ EventEmitter = NoEventEmitter{}
+
+// NoEventEmitter is usually used in the environment for script execution,
+// where no event should be emitted.
+type NoEventEmitter struct{}
+
+func (NoEventEmitter) EmitEvent(event cadence.Event) error {
+	return errors.NewOperationNotSupportedError("EmitEvent")
```

Comment:
Any chance to make this non-error? This error is limiting a lot of good stuff. ref: https://github.com/onflow/flow-go/issues/2813